File: sdk/macos/Sources/Xid/Xid.swift
// Xid.swift
// XID macOS Swift SDK
// MIT
//
// Main entry point. Provides configure / signIn / handleRedirect / getSession /
// getAccessToken / signOut / setTokenStorage, aligned with the platform matrix
// Shared native contract.

import AuthenticationServices
import Foundation

// MARK: - Xid main class

/// XID SDK main class. Access via Xid.shared singleton.
/// All public methods are thread-safe (internal state protected by NSLock).
public final class Xid: @unchecked Sendable {
    // MARK: - Singleton

    public static let shared = Xid()
    private init() {}

    // MARK: - Internal state (protected by stateLock)

    private let stateLock = NSLock()
    private var _config: XidConfiguration?
    private var _discoveryLoader: OIDCDiscoveryLoader?
    private let userInfoClient = UserInfoClient()
    private let endSessionClient = EndSessionClient()
    // var (not let) so tests can swap in a client backed by a mocked URLSession.
    var guestAuthClient = GuestAuthClient()

    private var config: XidConfiguration {
        get throws {
            stateLock.lock()
            defer { stateLock.unlock() }
            guard let c = _config else { throw XidError.notConfigured }
            return c
        }
    }

    private var discoveryLoader: OIDCDiscoveryLoader {
        get throws {
            stateLock.lock()
            defer { stateLock.unlock() }
            guard let d = _discoveryLoader else { throw XidError.notConfigured }
            return d
        }
    }

    // MARK: - Active authorization session

    @MainActor private var activeAuthSession: AuthorizationSession?

    // MARK: - Public API

    /// Initializes the SDK configuration. Must be called before any other method.
    ///
    /// Example:
    /// ```swift
    /// Xid.shared.configure(options: XidConfiguration(
    ///     issuer: URL(string: "https://xid.dev")!,
    ///     clientId: "your_client_id",
    ///     redirectUri: URL(string: "com.example.app://auth/callback")!
    /// ))
    /// ```
    public func configure(options: XidConfiguration) {
        stateLock.lock()
        defer { stateLock.unlock() }
        _config = options
        _discoveryLoader = OIDCDiscoveryLoader(issuer: options.issuer)
    }

    /// Replaces the token persistence adapter. Can be called any time after configure.
    public func setTokenStorage(_ adapter: TokenStorageAdapter) throws {
        _ = try config
        stateLock.lock()
        defer { stateLock.unlock() }
        _config?.tokenStorage = adapter
    }

    /// Starts the Hosted Auth sign-in flow.
    /// Opens a system browser session (/authorize). After completion, calls handleRedirect(url:).
    ///
    /// - Parameter options: Extra parameters, e.g. ["prompt": "login"] to force re-authentication.
    @MainActor
    public func signIn(options: [String: String] = [:]) async throws {
        let c = try config
        try c.validatePublicClientScopes()
        let discovery = try await discoveryLoader.load()

        // Generate PKCE pair
        let pkce = try PKCE()

        // Generate random state for CSRF protection
        var stateBytes = [UInt8](repeating: 0, count: 16)
        let stateStatus = SecRandomCopyBytes(kSecRandomDefault, stateBytes.count, &stateBytes)
        guard stateStatus == errSecSuccess else {
            throw XidError.pkceGenerationFailed("SecRandomCopyBytes state failed: OSStatus \(stateStatus)")
        }
        let state = Data(stateBytes).base64URLEncodedString()

        // Persist verifier and state in Keychain for the duration of the flow
        try c.tokenStorage.save(key: StorageKey.pkceVerifier, value: pkce.verifier)
        try c.tokenStorage.save(key: StorageKey.oauthState, value: state)

        // Build /authorize URL
        guard let authURL = AuthorizationURLBuilder.build(
            authorizationEndpoint: discovery.authorizationEndpoint,
            clientId: c.clientId,
            redirectUri: c.redirectUri,
            scopes: c.scopes,
            pkce: pkce,
            state: state,
            additionalParams: options
        ) else {
            throw XidError.discoveryFailed("Could not build authorization URL")
        }

        // Extract custom URL scheme, or nil for universal links
        let scheme: String? = {
            let s = c.redirectUri.scheme ?? ""
            return s.hasPrefix("http") ? nil : s
        }()

        let session = AuthorizationSession()
        activeAuthSession = session

        let callbackURL = try await session.start(
            authorizationURL: authURL,
            callbackScheme: scheme,
            prefersEphemeral: c.prefersEphemeralWebBrowserSession
        )
        activeAuthSession = nil

        _ = try await handleRedirect(url: callbackURL)
    }

    /// Handles a universal link or custom scheme callback URL.
    ///
    /// Call from the app's URL handling delegate:
    /// ```swift
    /// func application(_ app: NSApplication, open urls: [URL]) {
    ///     if let url = urls.first {
    ///         Task { try await Xid.shared.handleRedirect(url: url) }
    ///     }
    /// }
    /// ```
    ///
    /// - Returns: The XidSession built from the exchanged tokens.
    @discardableResult
    public func handleRedirect(url: URL) async throws -> XidSession {
        let c = try config
        try c.validatePublicClientScopes()
        let discovery = try await discoveryLoader.load()

        guard let components = URLComponents(url: url, resolvingAgainstBaseURL: true),
              let queryItems = components.queryItems
        else {
            throw XidError.invalidCallbackURL("Cannot parse callback URL: \(url)")
        }

        // Check for OAuth error response
        if let errorCode = queryItems.first(where: { $0.name == "error" })?.value {
            let desc = queryItems.first(where: { $0.name == "error_description" })?.value
            throw XidError.oauthError(errorCode, desc)
        }

        guard let code = queryItems.first(where: { $0.name == "code" })?.value else {
            throw XidError.invalidCallbackURL("Callback URL missing 'code' parameter")
        }
        guard let returnedState = queryItems.first(where: { $0.name == "state" })?.value else {
            throw XidError.invalidCallbackURL("Callback URL missing 'state' parameter")
        }

        // Validate state (CSRF protection)
        guard let savedState = try c.tokenStorage.load(key: StorageKey.oauthState),
              savedState == returnedState
        else {
            throw XidError.stateMismatch
        }
        try c.tokenStorage.delete(key: StorageKey.oauthState)

        // Retrieve code_verifier
        guard let verifier = try c.tokenStorage.load(key: StorageKey.pkceVerifier) else {
            throw XidError.invalidCallbackURL("PKCE verifier not found; flow may have expired")
        }
        try c.tokenStorage.delete(key: StorageKey.pkceVerifier)

        // Exchange code for tokens
        let tokenClient = TokenEndpointClient(
            tokenEndpoint: discovery.tokenEndpoint,
            clientId: c.clientId
        )
        let tokenResponse = try await tokenClient.exchangeCode(
            code: code,
            redirectUri: c.redirectUri,
            codeVerifier: verifier
        )

        return try await persistAndBuildSession(
            tokenResponse: tokenResponse,
            storage: c.tokenStorage,
            discovery: discovery
        )
    }

    /// Signs in anonymously (Firebase-style guest mode).
    ///
    /// Lazy semantics: if a valid session already exists locally (hosted or guest),
    /// it is returned immediately without any network request. Otherwise fetches
    /// a one-time capability from {issuer}/auth/config?intent=sign-up, POSTs
    /// {issuer}/auth/guest, fetches /v1/me, then persists the complete guest
    /// session. Guest sessions hold no tokens
    /// (`accessToken` / `idToken` are nil) and are bound to this device only.
    ///
    /// - Parameter turnstileToken: Turnstile token, only required when the server
    ///   enforces Turnstile on /auth/guest (native apps typically do not need it).
    /// - Returns: The guest XidSession; `session.user.isAnonymous` is true.
    @discardableResult
    public func signInAnonymously(turnstileToken: String? = nil) async throws -> XidSession {
        let c = try config

        if let session = try await loadSession(storage: c.tokenStorage) {
            return session
        }

        let sessionCookie = try await guestAuthClient.signIn(
            issuer: c.issuer,
            turnstileToken: turnstileToken
        )
        let user = try await guestAuthClient.fetchUser(issuer: c.issuer, sessionCookie: sessionCookie)

        let userData = try JSONEncoder().encode(user)
        guard let userJSON = String(data: userData, encoding: .utf8) else {
            throw XidError.tokenStorageError("Cannot encode guest user as UTF-8")
        }
        do {
            try c.tokenStorage.save(key: StorageKey.guestSessionCookie, value: sessionCookie)
            try c.tokenStorage.save(key: StorageKey.guestUser, value: userJSON)
        } catch {
            // The two-key adapter has no transaction; rollback prevents a half session.
            try? c.tokenStorage.delete(key: StorageKey.guestSessionCookie)
            try? c.tokenStorage.delete(key: StorageKey.guestUser)
            throw error
        }

        return XidSession(
            accessToken: nil,
            refreshToken: nil,
            idToken: nil,
            expiresAt: .distantFuture,
            user: user
        )
    }

    /// Returns the current session while its access token is valid.
    /// Expired token sessions are cleared and require a new authorization.
    /// Returns nil if the user is not signed in.
    public func getSession() async throws -> XidSession? {
        let c = try config
        guard let session = try await loadSession(storage: c.tokenStorage) else {
            return nil
        }
        if session.accessToken != nil && session.isExpired {
            try clearStoredTokens(storage: c.tokenStorage)
            return nil
        }
        return session
    }

    /// Returns an unexpired access token. An expired token requires a new authorization.
    ///
    /// - Parameter forceRefresh: DPoP refresh is not implemented; true clears the session and
    ///   requires a new authorization.
    public func getAccessToken(forceRefresh: Bool = false) async throws -> String {
        let c = try config

        if !forceRefresh,
           let session = try await loadSession(storage: c.tokenStorage),
           let accessToken = session.accessToken,
           !session.isExpired
        {
            return accessToken
        }

        try clearStoredTokens(storage: c.tokenStorage)
        throw XidError.noActiveSession
    }

    /// Signs out: clears local tokens, optionally calls /end_session for server-side logout.
    ///
    /// - Parameter callEndSession: Whether to call /end_session (RP-initiated logout).
    public func signOut(callEndSession: Bool = true) async throws {
        let c = try config

        if callEndSession {
            let idToken = try c.tokenStorage.load(key: StorageKey.idToken)
            if let idToken, let discovery = try? await discoveryLoader.load(),
               let endSessionEndpoint = discovery.endSessionEndpoint
            {
                try? await endSessionClient.endSession(
                    endSessionEndpoint: endSessionEndpoint,
                    idTokenHint: idToken,
                    postLogoutRedirectUri: c.postLogoutRedirectUri
                )
            }
        }

        try clearStoredTokens(storage: c.tokenStorage)
        // Drop the jar copy of the guest cookie too, or a later signInAnonymously
        // would silently renew the signed-out guest session server-side.
        guestAuthClient.clearCookies(for: c.issuer)
    }

    // MARK: - Internal helpers

    private func persistAndBuildSession(
        tokenResponse: TokenResponse,
        storage: TokenStorageAdapter,
        discovery: OIDCDiscoveryDocument
    ) async throws -> XidSession {
        let expiresAt = Date().addingTimeInterval(TimeInterval(tokenResponse.expiresIn))

        try storage.save(key: StorageKey.accessToken, value: tokenResponse.accessToken)
        try? storage.delete(key: StorageKey.refreshToken)
        if let idToken = tokenResponse.idToken {
            try storage.save(key: StorageKey.idToken, value: idToken)
        }
        let isoFormatter = ISO8601DateFormatter()
        try storage.save(key: StorageKey.expiresAt, value: isoFormatter.string(from: expiresAt))

        let storedIdToken = try storage.load(key: StorageKey.idToken)
        let idTokenStr = tokenResponse.idToken ?? storedIdToken ?? ""
        let user = try await resolveUser(
            idToken: idTokenStr.isEmpty ? nil : idTokenStr,
            accessToken: tokenResponse.accessToken,
            discovery: discovery
        )

        // A token session replaces any prior guest session on this device.
        try? storage.delete(key: StorageKey.guestSessionCookie)
        try? storage.delete(key: StorageKey.guestUser)
        return XidSession(
            accessToken: tokenResponse.accessToken,
            refreshToken: nil,
            idToken: idTokenStr.isEmpty ? nil : idTokenStr,
            expiresAt: expiresAt,
            user: user
        )
    }

    private func resolveUser(
        idToken: String?,
        accessToken: String,
        discovery: OIDCDiscoveryDocument
    ) async throws -> XidUser {
        if let idToken, !idToken.isEmpty {
            let c = try config
            let verifier = IDTokenVerifier(
                issuer: c.issuer,
                clientId: c.clientId,
                jwksUri: discovery.jwksUri
            )
            return try await verifier.verifyAndDecodeUser(idToken)
        }

        guard let endpoint = discovery.userinfoEndpoint ?? URL(string: discovery.issuer)?
            .appendingPathComponent("userinfo")
        else {
            throw XidError.userInfoFetchFailed("No id_token and discovery has no userinfo_endpoint")
        }

        return try await userInfoClient.fetchUser(
            accessToken: accessToken,
            userinfoEndpoint: endpoint
        )
    }

    private func loadSession(storage: TokenStorageAdapter) async throws -> XidSession? {
        guard let accessToken = try storage.load(key: StorageKey.accessToken),
              let expiresAtStr = try storage.load(key: StorageKey.expiresAt)
        else {
            return try loadGuestSession(storage: storage)
        }

        let isoFormatter = ISO8601DateFormatter()
        guard let expiresAt = isoFormatter.date(from: expiresAtStr) else {
            return nil
        }

        let idTokenStr = (try? storage.load(key: StorageKey.idToken)) ?? ""
        let user: XidUser
        if idTokenStr.isEmpty {
            user = XidUser(
                sub: "unknown", email: nil, emailVerified: nil, name: nil, picture: nil,
                provisionedBy: nil
            )
        } else {
            user = (try? IDTokenDecoder.decodeUser(idTokenStr)) ?? XidUser(
                sub: "unknown", email: nil, emailVerified: nil, name: nil, picture: nil,
                provisionedBy: nil
            )
        }

        return XidSession(
            accessToken: accessToken,
            refreshToken: nil,
            idToken: idTokenStr.isEmpty ? nil : idTokenStr,
            expiresAt: expiresAt,
            user: user
        )
    }

    private func loadGuestSession(storage: TokenStorageAdapter) throws -> XidSession? {
        guard let userJSON = try storage.load(key: StorageKey.guestUser),
              let userData = userJSON.data(using: .utf8),
              let user = try? JSONDecoder().decode(XidUser.self, from: userData)
        else {
            return nil
        }
        return XidSession(
            accessToken: nil,
            refreshToken: nil,
            idToken: nil,
            expiresAt: .distantFuture,
            user: user
        )
    }

    private func clearStoredTokens(storage: TokenStorageAdapter) throws {
        try? storage.delete(key: StorageKey.accessToken)
        try? storage.delete(key: StorageKey.refreshToken)
        try? storage.delete(key: StorageKey.idToken)
        try? storage.delete(key: StorageKey.expiresAt)
        try? storage.delete(key: StorageKey.pkceVerifier)
        try? storage.delete(key: StorageKey.oauthState)
        try? storage.delete(key: StorageKey.guestSessionCookie)
        try? storage.delete(key: StorageKey.guestUser)
    }
}

File: sdk/android/src/main/kotlin/dev/xid/sdk/Xid.kt
package dev.xid.sdk

import android.content.Context
import android.net.Uri
import dev.xid.sdk.browser.AuthSession
import dev.xid.sdk.browser.CallbackResult
import dev.xid.sdk.guest.GuestAuthManager
import dev.xid.sdk.model.GetAccessTokenOptions
import dev.xid.sdk.model.OidcDiscovery
import dev.xid.sdk.model.SignInAnonymouslyOptions
import dev.xid.sdk.model.SignInOptions
import dev.xid.sdk.model.XidConfig
import dev.xid.sdk.model.XidException
import dev.xid.sdk.model.XidGuestSession
import dev.xid.sdk.model.XidSession
import dev.xid.sdk.pkce.PkceGenerator
import dev.xid.sdk.storage.EncryptedPrefsStorage
import dev.xid.sdk.storage.StorageKeys
import dev.xid.sdk.storage.TokenStorageAdapter
import dev.xid.sdk.token.TokenManager
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import okhttp3.OkHttpClient
import okhttp3.Request
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import java.util.concurrent.TimeUnit

/**
 * XID Android SDK 入口。
 * 所有公开 API 均为挂起函数, 需在协程上下文中调用。
 *
 * 典型用法:
 * ```kotlin
 * // Application.onCreate 或 Activity/Fragment 中
 * Xid.configure(context, XidConfig(
 *     issuer = "https://xid.dev",
 *     clientId = "your_client_id",
 *     redirectUri = "https://yourapp.example.com/callback",
 * ))
 *
 * // 触发登录
 * lifecycleScope.launch {
 *     try {
 *         val session = Xid.signIn(activity)
 *         // 登录成功
 *     } catch (e: XidException.UserCancelled) {
 *         // 用户取消
 *     }
 * }
 *
 * // 在处理回调的 Activity 中(onNewIntent 或 onCreate)
 * lifecycleScope.launch {
 *     intent.data?.let { uri ->
 *         val session = Xid.handleRedirect(uri.toString())
 *     }
 * }
 * ```
 *
 * 注意:
 * - configure 必须在任何其他方法调用前完成。
 * - signIn 启动 Chrome Custom Tabs 后控制权转移给浏览器,
 *   回调通过 handleRedirect 在新 Intent 中处理。
 * - 本类是 object(单例), 不持有 Activity 引用, 避免内存泄漏。
 */
object Xid {

    private var config: XidConfig? = null
    private var storage: TokenStorageAdapter? = null
    private var discovery: OidcDiscovery? = null

    // 用于保护 discovery 缓存和 token 刷新的并发安全
    private val mutex = Mutex()

    private val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
    }

    private val httpClient = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(15, TimeUnit.SECONDS)
        .build()

    // ---------------------------------------------------------------------------
    // configure
    // ---------------------------------------------------------------------------

    /**
     * 初始化 SDK。必须在 Application.onCreate 或第一个 Activity.onCreate 调用。
     *
     * @param context  Android Application Context(不要传 Activity, 避免泄漏)。
     * @param config   SDK 配置选项([XidConfig])。
     */
    fun configure(context: Context, config: XidConfig) {
        this.config = config
        // 使用 Application Context 避免 Activity 内存泄漏
        this.storage = EncryptedPrefsStorage(
            context.applicationContext,
            requireBiometricUnlock = config.requireBiometricUnlock,
        )
        // discovery 延迟加载(第一次 signIn 或 getSession 时拉取)
        this.discovery = null
    }

    // ---------------------------------------------------------------------------
    // setTokenStorage
    // ---------------------------------------------------------------------------

    /**
     * 替换默认的 [EncryptedPrefsStorage] 为自定义存储适配器。
     * 必须在 [configure] 之后、[signIn] 之前调用。
     *
     * 典型用途: 单元测试中注入内存 mock、或集成第三方 Keystore 方案。
     *
     * @param adapter  自定义存储适配器, 实现 [TokenStorageAdapter]。
     */
    fun setTokenStorage(adapter: TokenStorageAdapter) {
        requireConfigured()
        this.storage = adapter
    }

    // ---------------------------------------------------------------------------
    // signIn
    // ---------------------------------------------------------------------------

    /**
     * 启动 OIDC Authorization Code + PKCE S256 登录流程。
     * 通过 Chrome Custom Tabs 打开 XID Hosted Auth 页面。
     *
     * 流程:
     * 1. 加载 OIDC discovery。
     * 2. 生成 PKCE code_verifier + code_challenge(S256)。
     * 3. 生成随机 state 防 CSRF。
     * 4. 启动 Chrome Custom Tabs 打开 /authorize。
     * 5. (异步)等待 App Links 或 custom scheme 回调。
     * 6. Activity 收到回调后调用 [handleRedirect]。
     *
     * 注意: 本方法在 Chrome Custom Tabs 启动后立即返回(协程挂起让给浏览器),
     * 实际 session 在 [handleRedirect] 中获取。
     *
     * @param context  当前 Activity Context(用于启动 Custom Tabs)。
     * @param options  可选 signIn 参数([SignInOptions])。
     */
    suspend fun signIn(context: Context, options: SignInOptions = SignInOptions()) {
        val cfg = requireConfigured()
        val disc = ensureDiscovery(cfg)
        val stor = storage!!

        val pkce = PkceGenerator.generate()
        val authSession = AuthSession(storage = stor, discovery = disc)
        authSession.launch(context = context, config = cfg, pkce = pkce, options = options)
        // 控制权转移到浏览器, 挂起函数此处返回
        // 实际 session 在 handleRedirect 回调中处理
    }

    // ---------------------------------------------------------------------------
    // signInAnonymously
    // ---------------------------------------------------------------------------

    /**
     * 匿名访客登录(Firebase 式 guest 模式)。
     *
     * 调用 POST {issuer}/auth/guest 建立服务端会话, 捕获 Set-Cookie 中的会话 cookie
     * 并持久化到安全存储, 再调 GET /v1/me 取回用户信息。
     *
     * 惰性语义: 本地已存在 guest 会话时直接返回, 不发任何网络请求。
     * guest 不签发 OAuth token, 返回 [XidGuestSession](cookie 会话), 与 OIDC 登录的
     * [XidSession] 分离; getSession/getAccessToken 不感知 guest 会话。
     *
     * 转正语义: guest 完成任一正式登录后 sub 不变, RP 数据自然延续; 若改登另一个
     * 既有账号, sub 会变, 调用方可对比前后 user.sub 判定。
     *
     * @param options  [SignInAnonymouslyOptions], 可携带 turnstileToken。
     * @return [XidGuestSession], user.isAnonymous 为 true。
     * @throws [XidException.GuestSignInFailed] 建号失败或 /v1/me 未返回用户。
     */
    suspend fun signInAnonymously(
        options: SignInAnonymouslyOptions = SignInAnonymouslyOptions(),
    ): XidGuestSession {
        val cfg = requireConfigured()
        val stor = storage!!

        // 串行化, 避免并发调用在本地无会话时重复建号
        return mutex.withLock {
            GuestAuthManager(storage = stor, issuer = cfg.issuer)
                .signInAnonymously(options.turnstileToken)
        }
    }

    // ---------------------------------------------------------------------------
    // handleRedirect
    // ---------------------------------------------------------------------------

    /**
     * 处理 App Links 或 custom scheme 的回调 URI。
     * 在 Activity.onNewIntent 或 onCreate 中调用(当 Intent.data 非空时)。
     *
     * @param url  回调 URI 字符串(来自 Intent.data.toString())。
     * @return 成功登录后的 [XidSession]。
     * @throws [XidException.StateMismatch]         state 不匹配(CSRF 防护)。
     * @throws [XidException.AuthorizationError]    授权失败(用户拒绝或参数错误)。
     * @throws [XidException.TokenExchangeFailed]   code 换 token 失败。
     */
    suspend fun handleRedirect(url: String): XidSession {
        val cfg = requireConfigured()
        val disc = ensureDiscovery(cfg)
        val stor = storage!!

        val uri = Uri.parse(url)
        val authSession = AuthSession(storage = stor, discovery = disc)
        val callbackResult = authSession.handleCallback(uri)

        return when (callbackResult) {
            is CallbackResult.Error -> throw callbackResult.exception
            is CallbackResult.Success -> {
                val tokenManager = TokenManager(storage = stor, discovery = disc)
                val session = tokenManager.exchangeCode(
                    code = callbackResult.code,
                    codeVerifier = callbackResult.codeVerifier,
                    clientId = cfg.clientId,
                    redirectUri = cfg.redirectUri,
                )
                // 正式登录成功后丢弃本地 guest 会话, 避免后续 signInAnonymously 复用旧身份
                GuestAuthManager.clear(stor)
                session
            }
        }
    }

    // ---------------------------------------------------------------------------
    // getSession
    // ---------------------------------------------------------------------------

    /**
     * 获取当前会话。如果 access token 过期且有 refresh token, 自动刷新。
     *
     * @return 当前 [XidSession], 或 null 表示未登录。
     */
    suspend fun getSession(): XidSession? {
        val cfg = config ?: return null
        val stor = storage ?: return null

        val disc = runCatching { ensureDiscovery(cfg) }.getOrNull() ?: return null
        val tokenManager = TokenManager(storage = stor, discovery = disc)

        val session = tokenManager.loadSession() ?: return null

        // access token 提前 30 秒视为过期, 避免边界情况
        val isExpired = session.accessTokenExpiresAt - System.currentTimeMillis() < 30_000L
        if (!isExpired) return session

        // 尝试刷新
        if (session.refreshToken == null) return null

        return runCatching {
            mutex.withLock { tokenManager.refresh(cfg.clientId) }
        }.getOrNull()
    }

    // ---------------------------------------------------------------------------
    // getAccessToken
    // ---------------------------------------------------------------------------

    /**
     * 获取当前有效的 access token(JWT 字符串)。
     * 如果 token 过期, 自动使用 refresh token 刷新。
     *
     * @param options  [GetAccessTokenOptions], 可强制刷新。
     * @return access token 字符串。
     * @throws [XidException.NoSession]          未登录。
     * @throws [XidException.TokenRefreshFailed] 刷新失败(refresh token 失效)。
     */
    suspend fun getAccessToken(options: GetAccessTokenOptions = GetAccessTokenOptions()): String {
        val cfg = requireConfigured()
        val stor = storage!!
        val disc = ensureDiscovery(cfg)
        val tokenManager = TokenManager(storage = stor, discovery = disc)

        val session = tokenManager.loadSession() ?: throw XidException.NoSession()

        val isExpired = session.accessTokenExpiresAt - System.currentTimeMillis() < 30_000L
        if (!options.forceRefresh && !isExpired) {
            return session.accessToken
        }

        if (session.refreshToken == null) throw XidException.NoSession()

        return mutex.withLock { tokenManager.refresh(cfg.clientId) }.accessToken
    }

    // ---------------------------------------------------------------------------
    // signOut
    // ---------------------------------------------------------------------------

    /**
     * 退出登录。清除本地所有持久化 token。
     *
     * 如果 OIDC discovery 包含 end_session_endpoint, 可选择打开浏览器完成
     * RP-initiated logout(清除 XID 侧的 SSO session)。
     *
     * @param context           Android Context(如需打开 end_session URL)。
     * @param openEndSession    是否通过 Chrome Custom Tabs 打开 end_session URL。
     *                          默认 false, 仅清除本地 token。
     */
    suspend fun signOut(context: Context? = null, openEndSession: Boolean = false) {
        val stor = storage ?: return
        val cfg = config

        // id_token_hint 必须在清除本地 token 之前读取
        val idTokenHint = stor.get(StorageKeys.ID_TOKEN)

        val tokenManager = cfg?.let {
            val disc = runCatching { ensureDiscovery(it) }.getOrNull()
            disc?.let { d -> TokenManager(storage = stor, discovery = d, clientId = it.clientId) }
        }

        tokenManager?.clearAll() ?: stor.clearAll()
        GuestAuthManager.clear(stor)

        // RP-initiated logout(可选): GET end_session + Custom Tabs
        if (openEndSession && context != null && cfg != null && idTokenHint != null) {
            val disc = runCatching { ensureDiscovery(cfg) }.getOrNull()
            disc?.endSessionEndpoint?.let { endSessionUrl ->
                val finalUri = Uri.parse(
                    buildEndSessionUrl(
                        endSessionEndpoint = endSessionUrl,
                        idTokenHint = idTokenHint,
                        postLogoutRedirectUri = cfg.postLogoutRedirectUri,
                    ),
                )
                val customTabsIntent = androidx.browser.customtabs.CustomTabsIntent.Builder()
                    .setShareState(androidx.browser.customtabs.CustomTabsIntent.SHARE_STATE_OFF)
                    .build()
                customTabsIntent.launchUrl(context, finalUri)
            }
        }
    }

    internal fun buildEndSessionUrl(
        endSessionEndpoint: String,
        idTokenHint: String,
        postLogoutRedirectUri: String?,
    ): String {
        val params = buildList {
            add("id_token_hint=${encodeQueryValue(idTokenHint)}")
            if (postLogoutRedirectUri != null) {
                add("post_logout_redirect_uri=${encodeQueryValue(postLogoutRedirectUri)}")
            }
        }
        val separator = if ('?' in endSessionEndpoint) '&' else '?'
        return endSessionEndpoint + separator + params.joinToString("&")
    }

    private fun encodeQueryValue(value: String): String =
        URLEncoder.encode(value, StandardCharsets.UTF_8).replace("+", "%20")

    // ---------------------------------------------------------------------------
    // 内部辅助
    // ---------------------------------------------------------------------------

    private fun requireConfigured(): XidConfig {
        return config ?: throw XidException.NotConfigured()
    }

    /**
     * 懒加载并缓存 OIDC discovery 文档。
     * 线程安全: 通过 mutex 确保只发起一次 discovery 请求。
     */
    private suspend fun ensureDiscovery(cfg: XidConfig): OidcDiscovery {
        discovery?.let { return it }

        return mutex.withLock {
            // 双重检查, 避免 mutex 等待期间其他协程已完成 discovery
            discovery?.let { return@withLock it }

            val fetched = fetchDiscovery(cfg.issuer)
            discovery = fetched
            fetched
        }
    }

    private suspend fun fetchDiscovery(issuer: String): OidcDiscovery = withContext(Dispatchers.IO) {
        // issuer 不带末尾斜杠, 拼接标准 discovery URL
        val url = "${issuer.trimEnd('/')}/.well-known/openid-configuration"

        val request = Request.Builder()
            .url(url)
            .get()
            .header("Accept", "application/json")
            .build()

        val body = runCatching {
            httpClient.newCall(request).execute().use { response ->
                if (!response.isSuccessful) {
                    throw XidException.DiscoveryFailed("HTTP ${response.code}")
                }
                response.body?.string() ?: throw XidException.DiscoveryFailed("响应体为空")
            }
        }.getOrElse { e ->
            throw XidException.DiscoveryFailed(e.message ?: "未知错误", e)
        }

        runCatching {
            json.decodeFromString<OidcDiscovery>(body)
        }.getOrElse { e ->
            throw XidException.DiscoveryFailed("解析 discovery 文档失败", e)
        }
    }
}
